libcudacxx/include/cuda/__numeric/narrow.h

```diff
@@ -0,0 +1,101 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the libcu++ Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___NUMERIC_NARROW_H
+#define _CUDA___NUMERIC_NARROW_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if !_CCCL_COMPILER(NVRTC)
+#  include <stdexcept>
+#endif // !_CCCL_COMPILER(NVRTC)
+
+#include <cuda/std/__exception/terminate.h>
+#include <cuda/std/__type_traits/is_arithmetic.h>
+#include <cuda/std/__type_traits/is_signed.h>
+#include <cuda/std/__utility/forward.h>
+
+#include <cuda/std/__cccl/prologue.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+//! Uses static_cast to cast a value \p __from to type \p _To. This function is intended to show that narrowing and a
+//! potential change of the value is intended. Modelled after `gsl::narrow_cast`. See also the C++ Core
+//! Guidelines <a href="https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-narrowing">ES.46</a> and <a
+//! href="https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named">ES.49</a>.
+template <class _To, class _From>
+[[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _To narrow_cast(_From&& __from) noexcept
+{
+  return static_cast<_To>(_CUDA_VSTD::forward<_From>(__from));
+}
+
+#if !_CCCL_COMPILER(NVRTC)
```

Comment:
This needs to be guarded by `_CCCL_HAS_EXCEPTIONS()`
```suggestion
#if _CCCL_HAS_EXCEPTIONS()
```